geth/peers/topicpool.go

```diff
@@ -0,0 +1,243 @@
+package peers
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/mclock"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/discv5"
+	"github.com/status-im/status-go/geth/params"
+)
+
+// NewTopicPool returns instance of TopicPool
+func NewTopicPool(topic discv5.Topic, limits params.Limits, slowSync, fastSync time.Duration) *TopicPool {
+	return &TopicPool{
+		topic:    topic,
+		limits:   limits,
+		slowSync: slowSync,
+		fastSync: fastSync,
+		peers:    map[discv5.NodeID]*peerInfo{},
+	}
+}
+
+// TopicPool manages peers for topic.
+type TopicPool struct {
+	topic    discv5.Topic
+	limits   params.Limits
+	slowSync time.Duration
+	fastSync time.Duration
+
+	quit    chan struct{}
+	running int32
+
+	mu         sync.RWMutex
+	discWG     sync.WaitGroup
+	consumerWG sync.WaitGroup
+	connected  int
+	peers      map[discv5.NodeID]*peerInfo
+	period     chan time.Duration
+
+	cache *Cache
+}
+
+// SearchRunning returns true if search is running
+func (t *TopicPool) SearchRunning() bool {
+	return atomic.LoadInt32(&t.running) == 1
+}
+
+// MaxReached returns true if we connected with max number of peers.
+func (t *TopicPool) MaxReached() bool {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.connected == t.limits[1]
+}
+
+// ConfirmAdded called when peer was added by p2p Server.
+// 1. Skip a peer if it not in our peer table
+// 2. Add a peer to a cache.
+// 3. Disconnect a peer if it was connected after we reached max limit of peers.
+//    (we can't know in advance if peer will be connected, thats why we allow
+//     to overflow for short duration)
+// 4. Switch search to slow mode if it is running.
+func (t *TopicPool) ConfirmAdded(server *p2p.Server, nodeID discover.NodeID) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	// inbound connection
+	peer, exist := t.peers[discv5.NodeID(nodeID)]
+	if !exist {
+		return
+	}
+	// established connection means that the node is a viable candidate for a connection and can be cached
+	if t.cache != nil {
+		if err := t.cache.AddPeer(peer.node, t.topic); err != nil {
+			log.Error("failed to persist a peer", "error", err)
+		}
+	}
+	// when max limit is reached drop every peer after
+	if t.connected == t.limits[1] {
```

Comment:
maybe have `limits` as a structure with `Min` and `Max` fields for the readability sake?

Reply:
yeah, i already noticed that it would be clear, but i prefer not to change it in followup

Reply:
sure, works for me